metrics/replicacpumem.go

```diff
@@ -0,0 +1,83 @@
+package metrics
+
+import (
+	"time"
+
+	"github.com/relab/hotstuff/metrics/types"
+	"github.com/relab/hotstuff/modules"
+	"github.com/shirou/gopsutil/v3/cpu"
+	"github.com/shirou/gopsutil/v3/mem"
+)
+
+// CPUMem metics measures the percentage of cpu and memory utilization on the node.
+// If multiple replicas are run on the same node, then the data may be duplicated.
+// This is not enabled by default, to enable this metric add "cpumem" string to --metrics option.
+// Since it can interfere with the performance of the protocol, do not enable this metics unless required.
+// Interval for measuring the cpu and memory utilization should be above 100 milliseconds, for valid data collection.
+// This limitation is due to the gopsutil package.
+func init() {
+	RegisterReplicaMetric("cpumem", func() interface{} {
+		return &CPUMemStat{}
+	})
+	RegisterClientMetric("cpumem", func() interface{} {
+		return &CPUMemStat{}
+	})
+}
+
+// CPUMemStat measures CPU usage and Memory usage and record in the metric logs.
+type CPUMemStat struct {
+	mods *modules.Modules
+}
+
+// InitModule gives the module access to the other modules.
+func (c *CPUMemStat) InitModule(mods *modules.Modules) {
+	c.mods = mods
+	c.mods.EventLoop().RegisterObserver(types.TickEvent{}, func(event interface{}) {
+		c.tick(event.(types.TickEvent))
+	})
+	c.mods.Logger().Info("CPU-Memory stats metric enabled")
+	// Percent with 0 interval returns 0 usage when called first time.
```

Comment:
I think this comment should be more descriptive. Maybe:
```
The cpu.Percent function returns the CPU usage since the last call when called with an interval of 0.
This initial call ensures that the first measurement of the CPU usage is nonzero.
```